zerver/lib/thumbnail.py

```diff
@@ -1,34 +1,111 @@
-import io
-from typing import Optional, Tuple
+import logging
+import os
+import re
+from contextlib import contextmanager
+from dataclasses import dataclass
+from typing import Dict, Iterator, Optional, Tuple, Type, TypeVar
 from urllib.parse import urljoin
 
+import pyvips
 from django.utils.http import url_has_allowed_host_and_scheme
 from django.utils.translation import gettext as _
-from PIL import GifImagePlugin, Image, ImageOps, PngImagePlugin
-from PIL.Image import DecompressionBombError
+from typing_extensions import override
 
 from zerver.lib.camo import get_camo_url
 from zerver.lib.exceptions import ErrorCode, JsonableError
+from zerver.lib.queue import queue_event_on_commit
+from zerver.models import AbstractAttachment, ImageAttachment
 
 DEFAULT_AVATAR_SIZE = 100
 MEDIUM_AVATAR_SIZE = 500
 DEFAULT_EMOJI_SIZE = 64
 
-# These sizes were selected based on looking at the maximum common
-# sizes in a library of animated custom emoji, balanced against the
-# network cost of very large emoji images.
-MAX_EMOJI_GIF_SIZE = 128
+# We refuse to deal with any image whose total pixelcount exceeds this.
+IMAGE_BOMB_TOTAL_PIXELS = 90000000
+
+# Reject uploaded emoji which are larger than this
 MAX_EMOJI_GIF_FILE_SIZE_BYTES = 128 * 1024 * 1024  # 128 kb
```

Comment:
This is pre-existing, but there's something wrong with this line — the number says 128 MiB and the comment says "128 kb".

Reply:
Great catch, thanks!

Reply:
So this has always actually been "128MB" (since it was added in 1e65cdd0729b1eaf85db6619bb6d9a059b43e7c2), and as such useless, since nginx limits to much less than that. 
 The top 33 (out of 190) emoji in chat.zulip.org are all larger than 128k -- so if we keep this, we need to not make the limit retroactive.
 
 From the point of view of bandiwdth, it makes sense that we should have some cap on how big one of the emoji is -- but we care about the size _after_ thumbnailing, not before.  Only 3/190 of the post-thumbnailed emoji are > 128k, so this seems like a reasonable cutoff to impose on post-thumbnailed images.

Reply:
Yeah, that's a reasonable strategy. Though are those 3 exceptions animated emoji? One could imagine just enforcing 128k as the still-frame size for those.